src/sqlfluff/dialects/dialect_ansi.py

```diff
@@ -1479,6 +1485,16 @@ class CaseExpressionSegment(BaseSegment):
             Sequence(
                 Ref("SimpleArrayTypeGrammar", optional=True), Ref("ArrayLiteralSegment")
             ),
+            Sequence(
+                Ref("DatatypeSegment"),
+                OneOf(
+                    Ref("QuotedLiteralSegment"),
+                    Ref("NumericLiteralSegment"),
+                    Ref("BooleanLiteralGrammar"),
+                    Ref("NullLiteralSegment"),
+                    Ref("DateTimeLiteralGrammar"),
+                ),
+            ),
```

Comment:
This is the bit that's changing the `NOT` datatype. Does this need to be this generic? Changing to the following avoids the issue for example:

```suggestion
            Sequence(
                Ref.keyword("TIMESTAMP"),
                OneOf(
                    Ref("QuotedLiteralSegment"),
                    Ref("NumericLiteralSegment"),
                    Ref("BooleanLiteralGrammar"),
                    Ref("NullLiteralSegment"),
                    Ref("DateTimeLiteralGrammar"),
                ),
            ),
```

Reply:
It does need to be generic, because it's capturing expression syntax like:
```
INTEGER 1 + my_column
```

Reply:
Yeah my “fix” is nothing of the sort 🤦‍♂️